File: test/transpiler/pheno_pkg/src/cs/registerzadok.cs
using System;
using System.Collections.Generic;
public class Registerzadok_
{
    public static Tuple<int,string,List<string>,List<string>,List<double>>  registerzadok_(double cumulTT,double phase,double leafNumber,List<string> calendarMoments,List<string> calendarDates,List<double> calendarCumuls,double cumulTTFromZC_65,string currentdate,double der,double slopeTSFLN,double intTSFLN,double finalLeafNumber,string currentZadokStage,int hasZadokStageChanged)
    {
        //- Description:
    //            - Model Name: RegisterZadok Model
    //            - Author: Pierre MARTRE
    //            - Reference: Modeling development phase in the 
    //                Wheat Simulation Model SiriusQuality.
    //                See documentation at http://www1.clermont.inra.fr/siriusquality/?page_id=427
    //            - Institution: INRA/LEPSE Montpellier
    //            - Abstract: Record the zadok stage in the calendar
    //    	
        //- inputs:
    //            - name: cumulTT
    //                          - min : -200
    //                          - default : 354.582294511779
    //                          - max : 10000
    //                          - uri : some url
    //                          - variablecategory : auxiliary
    //                          - datatype : DOUBLE
    //                          - inputtype : variable
    //                          - unit : °C d
    //                          - description : 
    //            - name: phase
    //                          - min : 0
    //                          - default : 2
    //                          - max : 7
    //                          - uri : some url
    //                          - variablecategory : state
    //                          - datatype : DOUBLE
    //                          - inputtype : variable
    //                          - unit : 
    //                          - description : instance of the phase class . You can get the name of the phase using phase.getPhaseAsString(PhaseValue) 
    //            - name: leafNumber
    //                          - min : 0
    //                          - default : 4.8854219661087575
    //                          - max : 25
    //                          - uri : some url
    //                          - variablecategory : state
    //                          - datatype : DOUBLE
    //                          - inputtype : variable
    //                          - unit : leaf
    //                          - description : Actual number of phytomers
    //            - name: calendarMoments
    //                          - variablecategory : auxiliary
    //                          - datatype : STRINGLIST
    //                          - default : ['Sowing']
    //                          - inputtype : variable
    //                          - unit : 
    //                          - description : List containing apparition of each stage
    //            - name: calendarDates
    //                          - variablecategory : auxiliary
    //                          - datatype : DATELIST
    //                          - default : ['21/3/2007']
    //                          - inputtype : variable
    //                          - unit : 
    //                          - description : List containing  the dates of the wheat developmental phases
    //            - name: calendarCumuls
    //                          - variablecategory : auxiliary
    //                          - datatype : DOUBLELIST
    //                          - default : [0.0]
    //                          - inputtype : variable
    //                          - unit : °C d
    //                          - description : list containing for each stage occured its cumulated thermal times
    //            - name: cumulTTFromZC_65
    //                          - min : -200
    //                          - default : 0
    //                          - max : 10000
    //                          - uri : some url
    //                          - variablecategory : auxiliary
    //                          - datatype : DOUBLE
    //                          - inputtype : variable
    //                          - unit : °C d
    //                          - description : cumul of the thermal time (DeltaTT) since the moment ZC_65
    //            - name: currentdate
    //                          - min : 
    //                          - default : 9/4/2007
    //                          - max : 
    //                          - uri : some url
    //                          - variablecategory : auxiliary
    //                          - datatype : DATE
    //                          - inputtype : variable
    //                          - unit : 
    //                          - description : current date
    //            - name: der
    //                          - parametercategory : species
    //                          - min : 0
    //                          - datatype : DOUBLE
    //                          - max : 10000
    //                          - uri : some url
    //                          - default : 300.0
    //                          - inputtype : parameter
    //                          - unit : °C d
    //                          - description : Duration of the endosperm endoreduplication phase
    //            - name: slopeTSFLN
    //                          - parametercategory : species
    //                          - min : 0
    //                          - datatype : DOUBLE
    //                          - max : 10000
    //                          - uri : some url
    //                          - default : 0.9
    //                          - inputtype : parameter
    //                          - unit : 
    //                          - description : used to calculate Terminal spikelet
    //            - name: intTSFLN
    //                          - parametercategory : species
    //                          - min : 0
    //                          - datatype : DOUBLE
    //                          - max : 10000
    //                          - uri : some url
    //                          - default : 0.9
    //                          - inputtype : parameter
    //                          - unit : 
    //                          - description : used to calculate Terminal spikelet
    //            - name: finalLeafNumber
    //                          - min : 0
    //                          - default : 8.797582013199484
    //                          - max : 10000
    //                          - uri : some url
    //                          - variablecategory : state
    //                          - datatype : DOUBLE
    //                          - inputtype : variable
    //                          - unit : leaf
    //                          - description : final leaf number
    //            - name: currentZadokStage
    //                          - min : 
    //                          - datatype : STRING
    //                          - max : 
    //                          - uri : some url
    //                          - default : MainShootPlus1Tiller
    //                          - inputtype : variable
    //                          - unit : 
    //                          - description : current zadok stage
    //            - name: hasZadokStageChanged
    //                          - min : 0
    //                          - default : 0
    //                          - max : 1
    //                          - uri : some url
    //                          - variablecategory : state
    //                          - datatype : INT
    //                          - inputtype : variable
    //                          - unit : 
    //                          - description : true if the zadok stage has changed this time step
        //- outputs:
    //            - name: hasZadokStageChanged
    //                          - min : 0
    //                          - variablecategory : state
    //                          - max : 1
    //                          - uri : some url
    //                          - datatype : INT
    //                          - unit : 
    //                          - description : true if the zadok stage has changed this time step
    //            - name: currentZadokStage
    //                          - datatype : STRING
    //                          - variablecategory : auxiliary
    //                          - uri : some url
    //                          - unit :  
    //                          - description : current zadok stage
    //            - name: calendarMoments
    //                          - variablecategory : auxiliary
    //                          - datatype : STRINGLIST
    //                          - unit : 
    //                          - description :  List containing apparition of each stage
    //            - name: calendarDates
    //                          - variablecategory : auxiliary
    //                          - datatype : DATELIST
    //                          - unit : 
    //                          - description :  List containing  the dates of the wheat developmental phases
    //            - name: calendarCumuls
    //                          - variablecategory : auxiliary
    //                          - datatype : DOUBLELIST
    //                          - unit : °C d
    //                          - description :  list containing for each stage occured its cumulated thermal times
        int roundedFinalLeafNumber;
        roundedFinalLeafNumber = (int)(finalLeafNumber + 0.5d);
        if (leafNumber >= 4.0d && !calendarMoments.Contains("MainShootPlus1Tiller"))
        {
            calendarMoments.Add("MainShootPlus1Tiller");
            calendarCumuls.Add(cumulTT);
            calendarDates.Add(currentdate);
            hasZadokStageChanged = 1;
            currentZadokStage = "MainShootPlus1Tiller";
        }
        else if ( leafNumber >= 5.0d && !calendarMoments.Contains("MainShootPlus2Tiller"))
        {
            calendarMoments.Add("MainShootPlus2Tiller");
            calendarCumuls.Add(cumulTT);
            calendarDates.Add(currentdate);
            hasZadokStageChanged = 1;
            currentZadokStage = "MainShootPlus2Tiller";
        }
        else if ( leafNumber >= 6.0d && !calendarMoments.Contains("MainShootPlus3Tiller"))
        {
            calendarMoments.Add("MainShootPlus3Tiller");
            calendarCumuls.Add(cumulTT);
            calendarDates.Add(currentdate);
            hasZadokStageChanged = 1;
            currentZadokStage = "MainShootPlus3Tiller";
        }
        else if ( finalLeafNumber > 0.0d && leafNumber >= slopeTSFLN * finalLeafNumber - intTSFLN && !calendarMoments.Contains("TerminalSpikelet"))
        {
            calendarMoments.Add("TerminalSpikelet");
            calendarCumuls.Add(cumulTT);
            calendarDates.Add(currentdate);
            hasZadokStageChanged = 1;
            currentZadokStage = "TerminalSpikelet";
        }
        else if ( leafNumber >= roundedFinalLeafNumber - 4.0d && roundedFinalLeafNumber - 4 > 0 && !calendarMoments.Contains("PseudoStemErection"))
        {
            calendarMoments.Add("PseudoStemErection");
            calendarCumuls.Add(cumulTT);
            calendarDates.Add(currentdate);
            hasZadokStageChanged = 1;
            currentZadokStage = "PseudoStemErection";
        }
        else if ( leafNumber >= roundedFinalLeafNumber - 3.0d && roundedFinalLeafNumber - 3 > 0 && !calendarMoments.Contains("1stNodeDetectable"))
        {
            calendarMoments.Add("1stNodeDetectable");
            calendarCumuls.Add(cumulTT);
            calendarDates.Add(currentdate);
            hasZadokStageChanged = 1;
            currentZadokStage = "1stNodeDetectable";
        }
        else if ( leafNumber >= roundedFinalLeafNumber - 2.0d && roundedFinalLeafNumber - 2 > 0 && !calendarMoments.Contains("2ndNodeDetectable"))
        {
            calendarMoments.Add("2ndNodeDetectable");
            calendarCumuls.Add(cumulTT);
            calendarDates.Add(currentdate);
            hasZadokStageChanged = 1;
            currentZadokStage = "2ndNodeDetectable";
        }
        else if ( leafNumber >= roundedFinalLeafNumber - 1.0d && roundedFinalLeafNumber - 1 > 0 && !calendarMoments.Contains("FlagLeafJustVisible"))
        {
            calendarMoments.Add("FlagLeafJustVisible");
            calendarCumuls.Add(cumulTT);
            calendarDates.Add(currentdate);
            hasZadokStageChanged = 1;
            currentZadokStage = "FlagLeafJustVisible";
        }
        else if ( !calendarMoments.Contains("MidGrainFilling") && phase == 4.5d && cumulTTFromZC_65 >= der)
        {
            calendarMoments.Add("MidGrainFilling");
            calendarCumuls.Add(cumulTT);
            calendarDates.Add(currentdate);
            hasZadokStageChanged = 1;
            currentZadokStage = "MidGrainFilling";
        }
        else
        {
            hasZadokStageChanged = 0;
        }
        return Tuple.Create(hasZadokStageChanged, currentZadokStage, calendarMoments, calendarDates, calendarCumuls);
    }
}

File: test/Models/pheno_pkg/src/cpp/Registerzadok.cpp
#define _USE_MATH_DEFINES
#include <cmath>
#include <iostream>
# include<vector>
# include<string>
# include<numeric>
# include<algorithm>
# include<array>
#include <map>
# include <tuple>
#include "Registerzadok.h"
using namespace std;

Registerzadok::Registerzadok() { }
float Registerzadok::getder() {return this-> der; }
float Registerzadok::getslopeTSFLN() {return this-> slopeTSFLN; }
float Registerzadok::getintTSFLN() {return this-> intTSFLN; }
string Registerzadok::getsowingDate() {return this-> sowingDate; }
void Registerzadok::setder(float _der) { this->der = _der; }
void Registerzadok::setslopeTSFLN(float _slopeTSFLN) { this->slopeTSFLN = _slopeTSFLN; }
void Registerzadok::setintTSFLN(float _intTSFLN) { this->intTSFLN = _intTSFLN; }
void Registerzadok::setsowingDate(string _sowingDate) { this->sowingDate = _sowingDate; }
void Registerzadok::Calculate_Model(PhenologyState& s, PhenologyState& s1, PhenologyRate& r, PhenologyAuxiliary& a)
{
    //- Name: RegisterZadok -Version: 1.0, -Time step: 1
    //- Description:
    //            * Title: RegisterZadok Model
    //            * Author: Pierre MARTRE
    //            * Reference: Modeling development phase in the 
    //                Wheat Simulation Model SiriusQuality.
    //                See documentation at http://www1.clermont.inra.fr/siriusquality/?page_id=427
    //            * Institution: INRA/LEPSE Montpellier
    //            * Abstract: Record the zadok stage in the calendar
    //    	
    //- inputs:
    //            * name: cumulTT
    //                          ** description : 
    //                          ** variablecategory : auxiliary
    //                          ** datatype : DOUBLE
    //                          ** min : -200
    //                          ** max : 10000
    //                          ** default : 354.582294511779
    //                          ** unit : °C d
    //                          ** uri : some url
    //                          ** inputtype : variable
    //            * name: phase
    //                          ** description : instance of the phase class . You can get the name of the phase using phase.getPhaseAsString(PhaseValue) 
    //                          ** variablecategory : state
    //                          ** inputtype : variable
    //                          ** datatype : DOUBLE
    //                          ** min : 0
    //                          ** max : 7
    //                          ** default : 2
    //                          ** unit : 
    //                          ** uri : some url
    //            * name: leafNumber
    //                          ** description : Actual number of phytomers
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** min : 0
    //                          ** max : 25
    //                          ** default : 4.8854219661087575
    //                          ** unit : leaf
    //                          ** uri : some url
    //                          ** inputtype : variable
    //            * name: calendarMoments
    //                          ** description : List containing apparition of each stage
    //                          ** variablecategory : state
    //                          ** datatype : STRINGLIST
    //                          ** default : ['Sowing']
    //                          ** unit : 
    //                          ** inputtype : variable
    //            * name: calendarDates
    //                          ** description : List containing  the dates of the wheat developmental phases
    //                          ** variablecategory : state
    //                          ** datatype : DATELIST
    //                          ** default : ['2007/3/21']
    //                          ** unit : 
    //                          ** inputtype : variable
    //            * name: calendarCumuls
    //                          ** description : list containing for each stage occured its cumulated thermal times
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLELIST
    //                          ** default : [0.0]
    //                          ** unit : °C d
    //                          ** inputtype : variable
    //            * name: cumulTTFromZC_65
    //                          ** description : cumul of the thermal time (DeltaTT) since the moment ZC_65
    //                          ** variablecategory : auxiliary
    //                          ** datatype : DOUBLE
    //                          ** min : -200
    //                          ** max : 10000
    //                          ** default : 0
    //                          ** unit : °C d
    //                          ** uri : some url
    //                          ** inputtype : variable
    //            * name: currentdate
    //                          ** description : current date
    //                          ** variablecategory : auxiliary
    //                          ** datatype : DATE
    //                          ** min : 
    //                          ** max : 
    //                          ** default : 2007/4/9
    //                          ** unit : 
    //                          ** uri : some url
    //                          ** inputtype : variable
    //            * name: der
    //                          ** description : Duration of the endosperm endoreduplication phase
    //                          ** parametercategory : species
    //                          ** datatype : DOUBLE
    //                          ** min : 0
    //                          ** max : 10000
    //                          ** default : 300.0
    //                          ** unit : °C d
    //                          ** uri : some url
    //                          ** inputtype : parameter
    //            * name: slopeTSFLN
    //                          ** description : used to calculate Terminal spikelet
    //                          ** parametercategory : species
    //                          ** datatype : DOUBLE
    //                          ** min : 0
    //                          ** max : 10000
    //                          ** default : 0.9
    //                          ** unit : 
    //                          ** uri : some url
    //                          ** inputtype : parameter
    //            * name: intTSFLN
    //                          ** description : used to calculate Terminal spikelet
    //                          ** parametercategory : species
    //                          ** datatype : DOUBLE
    //                          ** min : 0
    //                          ** max : 10000
    //                          ** default : 0.9
    //                          ** unit : 
    //                          ** uri : some url
    //                          ** inputtype : parameter
    //            * name: finalLeafNumber
    //                          ** description : final leaf number
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** min : 0
    //                          ** max : 10000
    //                          ** default : 8.797582013199484
    //                          ** unit : leaf
    //                          ** uri : some url
    //                          ** inputtype : variable
    //            * name: currentZadokStage
    //                          ** description : current zadok stage
    //                          ** variablecategory : state
    //                          ** datatype : STRING
    //                          ** min : 
    //                          ** max : 
    //                          ** default : MainShootPlus1Tiller
    //                          ** unit : 
    //                          ** uri : some url
    //                          ** inputtype : variable
    //            * name: hasZadokStageChanged_t1
    //                          ** description : true if the zadok stage has changed this time step
    //                          ** variablecategory : state
    //                          ** datatype : INT
    //                          ** min : 0
    //                          ** max : 1
    //                          ** default : 0
    //                          ** unit : 
    //                          ** uri : some url
    //                          ** inputtype : variable
    //            * name: sowingDate
    //                          ** description :  Date of Sowing
    //                          ** parametercategory : constant
    //                          ** datatype : DATE
    //                          ** min : 
    //                          ** max : 
    //                          ** default : 2007/3/21
    //                          ** unit : 
    //                          ** inputtype : parameter
    //- outputs:
    //            * name: hasZadokStageChanged
    //                          ** description : true if the zadok stage has changed this time step
    //                          ** variablecategory : state
    //                          ** datatype : INT
    //                          ** min : 0
    //                          ** max : 1
    //                          ** unit : 
    //                          ** uri : some url
    //            * name: currentZadokStage
    //                          ** description : current zadok stage
    //                          ** variablecategory : state
    //                          ** datatype : STRING
    //                          ** unit :  
    //                          ** uri : some url
    //            * name: calendarMoments
    //                          ** description :  List containing apparition of each stage
    //                          ** variablecategory : state
    //                          ** datatype : STRINGLIST
    //                          ** unit : 
    //            * name: calendarDates
    //                          ** description :  List containing  the dates of the wheat developmental phases
    //                          ** variablecategory : state
    //                          ** datatype : DATELIST
    //                          ** unit : 
    //            * name: calendarCumuls
    //                          ** description :  list containing for each stage occured its cumulated thermal times
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLELIST
    //                          ** unit : °C d
    float cumulTT = a.getcumulTT();
    float phase = s.getphase();
    float leafNumber = s.getleafNumber();
    vector<string> calendarMoments = s.getcalendarMoments();
    vector<string> calendarDates = s.getcalendarDates();
    vector<float> calendarCumuls = s.getcalendarCumuls();
    float cumulTTFromZC_65 = a.getcumulTTFromZC_65();
    string currentdate = a.getcurrentdate();
    float finalLeafNumber = s.getfinalLeafNumber();
    string currentZadokStage = s.getcurrentZadokStage();
    int hasZadokStageChanged_t1 = s1.gethasZadokStageChanged();
    int hasZadokStageChanged;
    int roundedFinalLeafNumber;
    roundedFinalLeafNumber = int(finalLeafNumber + 0.5f);
    if (leafNumber >= 4.0f && !(find(calendarMoments.begin(), calendarMoments.end(), "MainShootPlus1Tiller") != calendarMoments.end()))
    {
        calendarMoments.push_back("MainShootPlus1Tiller");
        calendarCumuls.push_back(cumulTT);
        calendarDates.push_back(currentdate);
        hasZadokStageChanged = 1;
        currentZadokStage = "MainShootPlus1Tiller";
    }
    else if ( leafNumber >= 5.0f && !(find(calendarMoments.begin(), calendarMoments.end(), "MainShootPlus2Tiller") != calendarMoments.end()))
    {
        calendarMoments.push_back("MainShootPlus2Tiller");
        calendarCumuls.push_back(cumulTT);
        calendarDates.push_back(currentdate);
        hasZadokStageChanged = 1;
        currentZadokStage = "MainShootPlus2Tiller";
    }
    else if ( leafNumber >= 6.0f && !(find(calendarMoments.begin(), calendarMoments.end(), "MainShootPlus3Tiller") != calendarMoments.end()))
    {
        calendarMoments.push_back("MainShootPlus3Tiller");
        calendarCumuls.push_back(cumulTT);
        calendarDates.push_back(currentdate);
        hasZadokStageChanged = 1;
        currentZadokStage = "MainShootPlus3Tiller";
    }
    else if ( finalLeafNumber > 0.0f && (leafNumber >= slopeTSFLN * finalLeafNumber - intTSFLN && !(find(calendarMoments.begin(), calendarMoments.end(), "TerminalSpikelet") != calendarMoments.end())))
    {
        calendarMoments.push_back("TerminalSpikelet");
        calendarCumuls.push_back(cumulTT);
        calendarDates.push_back(currentdate);
        hasZadokStageChanged = 1;
        currentZadokStage = "TerminalSpikelet";
    }
    else if ( leafNumber >= roundedFinalLeafNumber - 4.0f && roundedFinalLeafNumber - 4 > 0 && !(find(calendarMoments.begin(), calendarMoments.end(), "PseudoStemErection") != calendarMoments.end()))
    {
        calendarMoments.push_back("PseudoStemErection");
        calendarCumuls.push_back(cumulTT);
        calendarDates.push_back(currentdate);
        hasZadokStageChanged = 1;
        currentZadokStage = "PseudoStemErection";
    }
    else if ( leafNumber >= roundedFinalLeafNumber - 3.0f && roundedFinalLeafNumber - 3 > 0 && !(find(calendarMoments.begin(), calendarMoments.end(), "1stNodeDetectable") != calendarMoments.end()))
    {
        calendarMoments.push_back("1stNodeDetectable");
        calendarCumuls.push_back(cumulTT);
        calendarDates.push_back(currentdate);
        hasZadokStageChanged = 1;
        currentZadokStage = "1stNodeDetectable";
    }
    else if ( leafNumber >= roundedFinalLeafNumber - 2.0f && roundedFinalLeafNumber - 2 > 0 && !(find(calendarMoments.begin(), calendarMoments.end(), "2ndNodeDetectable") != calendarMoments.end()))
    {
        calendarMoments.push_back("2ndNodeDetectable");
        calendarCumuls.push_back(cumulTT);
        calendarDates.push_back(currentdate);
        hasZadokStageChanged = 1;
        currentZadokStage = "2ndNodeDetectable";
    }
    else if ( leafNumber >= roundedFinalLeafNumber - 1.0f && roundedFinalLeafNumber - 1 > 0 && !(find(calendarMoments.begin(), calendarMoments.end(), "FlagLeafJustVisible") != calendarMoments.end()))
    {
        calendarMoments.push_back("FlagLeafJustVisible");
        calendarCumuls.push_back(cumulTT);
        calendarDates.push_back(currentdate);
        hasZadokStageChanged = 1;
        currentZadokStage = "FlagLeafJustVisible";
    }
    else if ( !(find(calendarMoments.begin(), calendarMoments.end(), "MidGrainFilling") != calendarMoments.end()) && (phase == 4.5f && cumulTTFromZC_65 >= der))
    {
        calendarMoments.push_back("MidGrainFilling");
        calendarCumuls.push_back(cumulTT);
        calendarDates.push_back(currentdate);
        hasZadokStageChanged = 1;
        currentZadokStage = "MidGrainFilling";
    }
    else
    {
        hasZadokStageChanged = 0;
    }
    s.setcalendarMoments(calendarMoments);
    s.setcalendarDates(calendarDates);
    s.setcalendarCumuls(calendarCumuls);
    s.setcurrentZadokStage(currentZadokStage);
    s.sethasZadokStageChanged(hasZadokStageChanged);
}